lib/logstash/plugin_mixins/rabbitmq_connection.rb

```diff
@@ -0,0 +1,149 @@
+
+        # The vhost to use. If you don't know what this is, leave the default.
+        config :vhost, :validate => :string, :default => "/"
+
+        # Enable or disable SSL
+        config :ssl, :validate => :boolean, :default => false
+
+        # Validate SSL certificate
+        config :verify_ssl, :validate => :boolean, :default => false
+
+        # Enable or disable logging
+        config :debug, :validate => :boolean, :default => false, :deprecated => "Use the logstash --debug flag for this instead."
+
+        # Set this to automatically recover from a broken connection. You almost certainly don't want to override this!!!
+        config :automatic_recovery, :validate => :boolean, :default => true
+
+        # Time in seconds to wait before retrying a connection
+        config :connect_retry_interval, :validate => :number, :default => 1
+
+        # Passive queue creation? Useful for checking queue existance without modifying server state
+        config :passive, :validate => :boolean, :default => false
+
+        # Extra queue arguments as an array.
+        # To make a RabbitMQ queue mirrored, use: `{"x-ha-policy" => "all"}`
+        config :arguments, :validate => :array, :default => {}
+
+      end
+
+      def conn_str
+        "amqp://#{@user}@#{@host}:#{@port}#{@vhost}"
+      end
+
+      def close_connection
+        @rabbitmq_connection_stopping = true
+        @hare_info.channel.close if channel_open?
+        @hare_info.connection.close if connection_open?
+      end
+
+      def rabbitmq_settings
+        return @rabbitmq_settings if @rabbitmq_settings
+
+        s = {
+          :vhost => @vhost,
+          :host  => @host,
+          :port  => @port,
+          :user  => @user,
+          :automatic_recovery => @automatic_recovery,
+          :pass => @password ? @password.value : "guest",
+        }
+        s[:tls] = @ssl if @ssl
+        @rabbitmq_settings = s
+      end
+
+      def connect!
+        @hare_info = connect() unless @hare_info # Don't duplicate the conn!
+      rescue MarchHare::Exception => e
+        @logger.error("RabbitMQ connection error, will retry.",
+                      :message => e.message,
+                      :exception => e.class.name,
+                      :backtrace => e.backtrace)
+
+        sleep_for_retry
+        retry
+      end
+
+      def channel_open?
+        @hare_info && @hare_info.channel && @hare_info.channel.open?
+      end
+
+      def connection_open?
+        @hare_info && @hare_info.connection && @hare_info.connection.open?
+      end
+
+      def connected?
+        return nil unless @hare_info && @hare_info.connection
+        @hare_info.connection.connected?
+      end
+
+      private
+
+      def declare_exchange!(channel, exchange, exchange_type, durable)
+        @logger.debug("Declaring an exchange", :name => exchange,
+                      :type => exchange_type, :durable => durable)
```

Comment:
There are for performance reason, research from @jordansissel a few years ago https://github.com/jordansissel/experiments/tree/master/ruby/logger-string-vs-block

So I am with you with clarity and since its rarely called, its called often the problem will cause more degradation of performance than the `if`. don't change it.